packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js

```diff
@@ -1,61 +1,98 @@
+    const [componentMap, setComponentMap] = useState(DEMO);
     const { apiBase } = useApolloClient();
+    const history = useHistory();
     const { pathname } = useLocation();
     const isMountedRef = useRef(false);
 
-    const routeData = componentMap.get(pathname);
+    // Keep track of whether we have been mounted yet.
+    // Note that we are not unmounted on page transitions.
+    useEffect(() => {
+        isMountedRef.current = true;
+
+        return () => {
+            isMountedRef.current = false;
+        };
+    }, []);
 
     // ask Magento for a RootComponent that matches the current pathname
     useEffect(() => {
-        // avoid asking if we already know the answer
-        const isKnown = componentMap.has(pathname);
+        // Avoid setting state if unmounted.
+        if (!isMountedRef.current) {
+            return;
+        }
 
-        // ask again following a prior failure
-        const isNotFound = isKnown && componentMap.get(pathname).id === -1;
-        const shouldReload = isNotFound && navigator.onLine;
+        const shouldFetch = data => {
+            // Ask if we don't have any data.
+            if (!data) return true;
 
-        if (!isKnown || shouldReload) {
-            getRouteComponent(apiBase, pathname).then(
-                ({ component, id, pathname, routeError }) => {
-                    // avoid setting state if unmounted
-                    if (!isMountedRef.current) {
-                        return;
-                    }
+            // Ask again following a prior failure.
+            if (data.isNotFound && navigator.onLine) {
+                return true;
+            }
+
+            return false;
+        };
 
+        // Get the data for this pathname from our Map in local state.
+        const routeData = componentMap.get(pathname);
+
+        // If we have data but it's a redirect, perform the redirect.
+        if (routeData && routeData.isRedirect) {
+            history.replace(routeData.relativeUrl);
+            return;
+        }
+
+        if (shouldFetch(routeData)) {
+            getRouteComponent(apiBase, pathname).then(
+                ({
+                    component,
+                    id,
+                    pathname,
+                    redirectCode,
+                    relativeUrl,
+                    routeError
+                }) => {
                     // add the pathname to the browser cache
                     addToCache(pathname);
 
-                    // then add the pathname and result to local state
+                    // Update our Map in local state for this path.
                     setComponentMap(prevMap => {
                         const nextMap = new Map(prevMap);
+
                         const nextValue = routeError
                             ? { hasError: true, routeError }
+                            : id === -1
+                            ? { isNotFound: true }
```

Comment:
Was this fixing a bug that's not jumping out? I assume it related to the offline retry, but kinda lost track of that in the refactor.

Reply:
Not fixing a bug, just improving readability for the logic after the lookup.  Basically all of the "inside knowledge" of what `-1` means, or "is this a redirect?" is here, so that when we retrieve this data we can just look at bools like `hasError`, `isNotFound`, `isRedirect`, etc.

Reply:
Made this a bit easier to read in https://github.com/magento/pwa-studio/pull/2504/commits/5be73656c08286410d34942110d9546cf67f28b0.